packages/firestore/src/util/async_queue.ts

```diff
@@ -93,11 +95,18 @@ export class AsyncQueue {
     );
   }
 
+  /**
+   * Waits for tasks that are scheduled for immediate execution and rejects any
+   * tasks that are pending with a non-zero delay.
+   */
   drain(): Promise<void> {
-    // TODO(mikelehen): This should perhaps also drain items that are queued to
-    // run in the future (perhaps by artificially running them early), but since
-    // no tests need that yet, I didn't bother for now.
-    assert(this.delayedOpCount === 0, "draining doesn't handle delayed ops.");
-    return this.schedule(() => Promise.resolve(undefined));
+    this.delayedOperations.forEach((deferred, handle) => {
+      window.clearTimeout(handle);
+      deferred.reject(
+        new FirestoreError(Code.CANCELLED, 'Cancelled during queue drain')
```

Comment:
These failures are potentially user visible so we need a better message about what can cause these to happen and how to avoid them if that's a problem.

If draining only happens during shutdown, the message can reflect that: `'Operation cancelled by shutdown'`.

Reply:
This is only used in tests, but I could see that we may want to call this during shutdown as well.